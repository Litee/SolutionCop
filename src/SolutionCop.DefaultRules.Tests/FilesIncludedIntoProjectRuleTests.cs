using System.IO;
using System.Xml.Linq;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Xunit;

namespace SolutionCop.DefaultRules.Tests
{
    [UseReporter(typeof (DiffReporter))]
    [UseApprovalSubdirectory("ApprovedResults")]
    public class FilesIncludedIntoProjectRuleTests : ProjectRuleTest
    {
        public FilesIncludedIntoProjectRuleTests() : base(new FilesIncludedIntoProjectRule())
        {
        }

        [Fact]
        public void Should_pass_if_all_files_are_included()
        {
            var xmlConfig = XElement.Parse("<FilesIncludedIntoProject/>");
            ShouldPassNormally(new FileInfo(@"..\..\Data\FilesIncludedIntoProject\AllFilesIncludedIntoProject.csproj").FullName, xmlConfig);
        }

        [Fact]
        public void Should_fail_if_file_is_not_included()
        {
            var xmlConfig = XElement.Parse("<FilesIncludedIntoProject/>");
            ShouldFailNormally(new FileInfo(@"..\..\Data\FilesIncludedIntoProject\FileNotIncludedIntoProject.csproj").FullName, xmlConfig);
        }

        [Fact]
        public void Should_pass_if_file_is_not_included_but_project_is_an_exception()
        {
            var xmlConfig = XElement.Parse(@"
<FilesIncludedIntoProject>
  <Exception><Project>SomeNonExistingProject.csproj</Project></Exception>
  <Exception><Project>FileNotIncludedIntoProject.csproj</Project></Exception>
</FilesIncludedIntoProject>");
            ShouldPassNormally(new FileInfo(@"..\..\Data\FilesIncludedIntoProject\FileNotIncludedIntoProject.csproj").FullName, xmlConfig);
        }

        [Fact]
        public void Should_fail_if_exception_misses_project()
        {
            var xmlConfig = XElement.Parse(@"
<FilesIncludedIntoProject>
  <Exception>Some text</Exception>
  <Exception><Project>FileNotIncludedIntoProject.csproj</Project></Exception>
</FilesIncludedIntoProject>");
            ShouldFailOnConfiguration(new FileInfo(@"..\..\Data\FilesIncludedIntoProject\FileNotIncludedIntoProject.csproj").FullName, xmlConfig);
        }

        [Fact]
        public void Should_pass_if_rule_is_disabled()
        {
            var xmlConfig = XElement.Parse("<FilesIncludedIntoProject enabled=\"false\"/>");
            ShouldPassAsDisabled(new FileInfo(@"..\..\Data\FilesIncludedIntoProject\FileNotIncludedIntoProject.csproj").FullName, xmlConfig);
        }
    }
}