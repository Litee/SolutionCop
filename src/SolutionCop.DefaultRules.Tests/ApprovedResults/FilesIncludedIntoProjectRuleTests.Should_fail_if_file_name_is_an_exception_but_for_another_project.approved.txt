Errors[0] = File is not referenced in project: Dir\My.File.cs
Errors[1] = File is not referenced in project: Dir\SubDir\My.File.cs
Errors[2] = File is not referenced in project: Properties\AssemblyInfo.cs
